File: packages/server/cmd/renewlet/media_resolver_config.go
package main

import (
	"encoding/json"
	"strings"

	appstatic "github.com/zhiyingzzhou/renewlet/packages/server/internal/static"
)

type mediaResolverConfig struct {
	BuiltInProviders []struct {
		Provider string `json:"provider"`
		CDNBase  string `json:"cdnBase"`
		GitHub   struct {
			Owner         string `json:"owner"`
			Repo          string `json:"repo"`
			Branch        string `json:"branch"`
			LatestRelease bool   `json:"latestRelease"`
		} `json:"github"`
		PreferredVariants []string `json:"preferredVariants"`
	} `json:"builtInProviders"`
	Auto struct {
		PlanSuffixWords []string `json:"planSuffixWords"`
	} `json:"auto"`
	Search struct {
		MinReducedQueryLength int      `json:"minReducedQueryLength"`
		ModifierSuffixWords   []string `json:"modifierSuffixWords"`
	} `json:"search"`
	CandidateGroups struct {
		SearchFaviconReserve int `json:"searchFaviconReserve"`
	} `json:"candidateGroups"`
	Limits struct {
		DefaultCandidates   int `json:"defaultCandidates"`
		MaxCandidates       int `json:"maxCandidates"`
		MaxItems            int `json:"maxItems"`
		MaxCandidateDomains int `json:"maxCandidateDomains"`
	} `json:"limits"`
	RateLimit struct {
		DefaultMaxRequests int `json:"defaultMaxRequests"`
		DefaultWindowMs    int `json:"defaultWindowMs"`
	} `json:"rateLimit"`
	Scores struct {
		Exact           float64 `json:"exact"`
		Prefix          float64 `json:"prefix"`
		Contains        float64 `json:"contains"`
		AllParts        float64 `json:"allParts"`
		Subsequence     float64 `json:"subsequence"`
		SlugExactBoost  float64 `json:"slugExactBoost"`
		SlugPrefixBoost float64 `json:"slugPrefixBoost"`
		StrongThreshold float64 `json:"strongThreshold"`
		MediumThreshold float64 `json:"mediumThreshold"`
	} `json:"scores"`
	Favicon struct {
		FallbackTLDs map[string][]string `json:"fallbackTlds"`
		Providers    []struct {
			Provider    string `json:"provider"`
			URLTemplate string `json:"urlTemplate"`
		} `json:"providers"`
		KnownDomains map[string]string `json:"knownDomains"`
	} `json:"favicon"`
}

var mediaResolverCfg = loadMediaResolverConfig()

func loadMediaResolverConfig() mediaResolverConfig {
	var config mediaResolverConfig
	if err := json.Unmarshal(appstatic.MediaResolverConfig, &config); err != nil {
		panic("invalid embedded media resolver config: " + err.Error())
	}
	// 配置是 Go/前端/shared 共用的候选排序事实源；启动时失败比悄悄降级到空候选更容易定位发布错误。
	if len(config.BuiltInProviders) == 0 || config.Search.MinReducedQueryLength <= 0 || len(config.Search.ModifierSuffixWords) == 0 || config.CandidateGroups.SearchFaviconReserve <= 0 || config.CandidateGroups.SearchFaviconReserve >= config.Limits.MaxCandidates || config.Limits.DefaultCandidates <= 0 || config.Limits.MaxCandidates <= 0 || len(config.Favicon.Providers) == 0 {
		panic("invalid embedded media resolver config")
	}
	return config
}

func mediaResolverBuiltInProviderRank() map[string]int {
	out := map[string]int{}
	for index, provider := range mediaResolverCfg.BuiltInProviders {
		if provider.Provider != "" {
			out[provider.Provider] = index
		}
	}
	return out
}

func mediaResolverBuiltInProviderBase(provider string) string {
	for _, item := range mediaResolverCfg.BuiltInProviders {
		if item.Provider == provider {
			return item.CDNBase
		}
	}
	return ""
}

func mediaResolverBuiltInProviderPinnedBase(provider string, ref string) string {
	for _, item := range mediaResolverCfg.BuiltInProviders {
		if item.Provider == provider && item.GitHub.Owner != "" && item.GitHub.Repo != "" && ref != "" {
			if !strings.Contains(item.CDNBase, "/gh/") {
				return item.CDNBase
			}
			return "https://testingcf.jsdelivr.net/gh/" + item.GitHub.Owner + "/" + item.GitHub.Repo + "@" + ref
		}
	}
	return mediaResolverBuiltInProviderBase(provider)
}

func mediaResolverPreferredVariants(provider string) []string {
	for _, item := range mediaResolverCfg.BuiltInProviders {
		if item.Provider == provider {
			return item.PreferredVariants
		}
	}
	return nil
}

func mediaSearchModifierSuffixWordSet() map[string]struct{} {
	out := map[string]struct{}{}
	for _, word := range mediaResolverCfg.Search.ModifierSuffixWords {
		if word != "" {
			out[word] = struct{}{}
		}
	}
	return out
}

func mediaPlanSuffixWordSet() map[string]struct{} {
	out := map[string]struct{}{}
	for _, word := range mediaResolverCfg.Auto.PlanSuffixWords {
		if word != "" {
			out[word] = struct{}{}
		}
	}
	return out
}

File: packages/shared/src/media-resolver-config.ts
/**
 * 媒体候选 resolver 配置由 JSON 数据驱动。
 *
 * 图标 provider、CDN、预算、降词和限流都必须在这里收敛，避免前端、Go embedded static 和 Worker 各自排序。
 */
import { z } from "zod";
import { BUILT_IN_ICON_PROVIDERS } from "./built-in-icons";
import mediaResolverConfigJson from "../data/media-resolver-config.json";

const mediaResolverConfigSchema = z.object({
  builtInProviders: z.array(z.object({
    provider: z.enum(BUILT_IN_ICON_PROVIDERS),
    cdnBase: z.string().url(),
    github: z.object({
      owner: z.string().min(1),
      repo: z.string().min(1),
      branch: z.string().min(1),
      latestRelease: z.boolean(),
    }).strict(),
    preferredVariants: z.array(z.string().min(1)).min(1),
  }).strict()).length(BUILT_IN_ICON_PROVIDERS.length).refine((providers) => {
    const seen = new Set(providers.map((item) => item.provider));
    return BUILT_IN_ICON_PROVIDERS.every((provider) => seen.has(provider));
  }, "内置图标 provider 配置不完整"),
  auto: z.object({
    planSuffixWords: z.array(z.string().min(1)).min(1),
  }).strict(),
  search: z.object({
    minReducedQueryLength: z.number().int().positive(),
    modifierSuffixWords: z.array(z.string().min(1)).min(1),
  }).strict(),
  candidateGroups: z.object({
    searchFaviconReserve: z.number().int().positive(),
  }).strict(),
  limits: z.object({
    defaultCandidates: z.number().int().positive(),
    maxCandidates: z.number().int().positive(),
    maxItems: z.number().int().positive(),
    maxCandidateDomains: z.number().int().positive(),
  }).strict(),
  rateLimit: z.object({
    defaultMaxRequests: z.number().int().positive(),
    defaultWindowMs: z.number().int().positive(),
  }).strict(),
  scores: z.object({
    exact: z.number().positive(),
    prefix: z.number().positive(),
    contains: z.number().positive(),
    allParts: z.number().positive(),
    subsequence: z.number().positive(),
    slugExactBoost: z.number().nonnegative(),
    slugPrefixBoost: z.number().nonnegative(),
    strongThreshold: z.number().positive(),
    mediumThreshold: z.number().positive(),
  }).strict(),
  favicon: z.object({
    fallbackTlds: z.object({
      logo: z.array(z.string().min(1)).min(1),
      icon: z.array(z.string().min(1)).min(1),
    }).strict(),
    providers: z.array(z.object({
      provider: z.string().min(1),
      urlTemplate: z.string().includes("{domain}"),
    }).strict()).min(1),
    knownDomains: z.record(z.string().min(1), z.string().min(1)),
  }).strict(),
}).strict().refine((config) => config.candidateGroups.searchFaviconReserve < config.limits.maxCandidates, "favicon 预留候选数必须小于最大候选数");

// media resolver 配置是 Docker/Cloudflare/脚本共用事实源；模块加载即校验，避免某个运行面带着坏配置启动。
export const mediaResolverConfig = mediaResolverConfigSchema.parse(mediaResolverConfigJson);

export type MediaResolverConfig = typeof mediaResolverConfig;
